CI failure: `error: 503 (`Unable to query remote tags`) => nimble should retry at least once

either nimble should retry on failure, on nim CI should retry on failure to avoid frequent flaky test failures caused by spurious 503 errors


## eg below
2020-05-10T09:46:24.5110975Z Error: unhandled exception: Unable to query remote tags for https://gitlab.com/define-private-public/stb_image-Nim.git. Git returned: fatal: unable to access 'https://gitlab.com/define-private-public/stb_image-Nim.git/': The requested URL returned error: 503

2020-05-10T09:53:50.4742150Z Error: unhandled exception: Unable to query remote tags for https://gitlab.com/define-private-public/stb_image-Nim.git. Git returned: fatal: unable to access 'https://gitlab.com/define-private-public/stb_image-Nim.git/': The requested URL returned error: 503


2020-05-10T10:09:57.5608090Z Error: unhandled exception: Unable to query remote tags for https://gitlab.com/define-private-public/stb_image-Nim.git. Git returned: fatal: unable to access 'https://gitlab.com/define-private-public/stb_image-Nim.git/': The requested URL returned error: 503

2020-05-28T06:35:09.1232590Z Error: unhandled exception: Unable to query remote tags for https://github.com/status-im/nim-chronos. Git returned: fatal: unable to access 'https://github.com/status-im/nim-chronos/': Failed to connect to github.com port 443: Operation timed out

```
2020-05-10T10:09:55.4239540Z FAILURE! total: 49 passed: 45 skipped: 0 failed: 4
2020-05-10T10:09:55.4246460Z Error: execution of an external program failed: '/Users/runner/runners/2.169.1/work/Nim/Nim/testament/testament cat nimble-packages-1'
2020-05-10T10:09:55.4333730Z FAILURE
2020-05-10T10:09:55.4343720Z ##[error]Process completed with exit code 1.
2020-05-10T10:09:55.4478410Z ##[group]Run nim c -r tools/ci_testresults.nim
2020-05-10T10:09:55.4478640Z [36;1mnim c -r tools/ci_testresults.nim[0m
2020-05-10T10:09:55.4861700Z shell: /bin/bash --noprofile --norc -e -o pipefail {0}
2020-05-10T10:09:55.4862120Z env:
2020-05-10T10:09:55.4862220Z   NIM_TEST_PACKAGES: 1
2020-05-10T10:09:55.4862300Z ##[endgroup]
2020-05-10T10:09:56.2028540Z Hint: used config file '/Users/runner/runners/2.169.1/work/Nim/Nim/config/nim.cfg' [Conf]
2020-05-10T10:09:56.2101330Z Hint: used config file '/Users/runner/runners/2.169.1/work/Nim/Nim/config/config.nims' [Conf]
2020-05-10T10:09:56.2107390Z Hint: system [Processing]
2020-05-10T10:09:56.2107840Z Hint: widestrs [Processing]
2020-05-10T10:09:56.2108110Z Hint: io [Processing]
2020-05-10T10:09:56.2108490Z Hint: ci_testresults [Processing]
2020-05-10T10:09:56.2108880Z Hint: os [Processing]
2020-05-10T10:09:56.2109280Z Hint: strutils [Processing]
2020-05-10T10:09:56.2109620Z Hint: parseutils [Processing]
2020-05-10T10:09:56.2110260Z Hint: math [Processing]
2020-05-10T10:09:56.2110680Z Hint: bitops [Processing]
2020-05-10T10:09:56.2111070Z Hint: macros [Processing]
2020-05-10T10:09:56.2641300Z Hint: algorithm [Processing]
2020-05-10T10:09:56.2641970Z Hint: unicode [Processing]
2020-05-10T10:09:56.5608260Z Hint: pathnorm [Processing]
2020-05-10T10:09:56.5617140Z Hint: osseps [Processing]
2020-05-10T10:09:56.5659390Z Hint: posix [Processing]
2020-05-10T10:09:56.5854400Z Hint: times [Processing]
2020-05-10T10:09:56.5866330Z Hint: options [Processing]
2020-05-10T10:09:56.5872440Z Hint: typetraits [Processing]
2020-05-10T10:09:56.7141300Z Hint: json [Processing]
2020-05-10T10:09:56.7147040Z Hint: hashes [Processing]
2020-05-10T10:09:56.7223650Z Hint: tables [Processing]
2020-05-10T10:09:56.7408500Z Hint: lexbase [Processing]
2020-05-10T10:09:56.7417250Z Hint: streams [Processing]
2020-05-10T10:09:56.7642090Z Hint: parsejson [Processing]
2020-05-10T10:09:56.8238650Z Hint: sets [Processing]
2020-05-10T10:09:56.8309990Z Hint: strformat [Processing]
2020-05-10T10:09:56.9280350Z CC: stdlib_assertions.nim
2020-05-10T10:09:56.9284270Z CC: stdlib_io.nim
2020-05-10T10:09:56.9287000Z CC: stdlib_system.nim
2020-05-10T10:09:56.9295430Z CC: stdlib_parseutils.nim
2020-05-10T10:09:56.9837670Z CC: stdlib_math.nim
2020-05-10T10:09:56.9942190Z CC: stdlib_unicode.nim
2020-05-10T10:09:57.0212790Z CC: stdlib_strutils.nim
2020-05-10T10:09:57.0219000Z CC: stdlib_posix.nim
2020-05-10T10:09:57.0587850Z CC: stdlib_times.nim
2020-05-10T10:09:57.0606340Z CC: stdlib_os.nim
2020-05-10T10:09:57.1022660Z CC: stdlib_hashes.nim
2020-05-10T10:09:57.1074560Z CC: stdlib_tables.nim
2020-05-10T10:09:57.1174250Z CC: stdlib_streams.nim
2020-05-10T10:09:57.1509440Z CC: stdlib_lexbase.nim
2020-05-10T10:09:57.1855560Z CC: stdlib_parsejson.nim
2020-05-10T10:09:57.1897370Z CC: stdlib_json.nim
2020-05-10T10:09:57.2136300Z CC: stdlib_sets.nim
2020-05-10T10:09:57.2225980Z CC: stdlib_strformat.nim
2020-05-10T10:09:57.2696890Z CC: ci_testresults.nim
2020-05-10T10:09:57.3453330Z Hint:  [Link]
2020-05-10T10:09:57.4762530Z Hint: 74119 LOC; 1.497 sec; 95.32MiB peakmem; Debug build; proj: /Users/runner/runners/2.169.1/work/Nim/Nim/tools/ci_testresults.nim; out: /Users/runner/runners/2.169.1/work/Nim/Nim/tools/ci_testresults [SuccessX]
2020-05-10T10:09:57.4763010Z Hint: /Users/runner/runners/2.169.1/work/Nim/Nim/tools/ci_testresults  [Exec]
2020-05-10T10:09:57.5521990Z Category: nimble-packages-1
2020-05-10T10:09:57.5522690Z Name: https://github.com/mratsim/Arraymancer C
2020-05-10T10:09:57.5524150Z Action: run
2020-05-10T10:09:57.5524380Z Result: reInstallFailed
2020-05-10T10:09:57.5525080Z -------- Expected -------
2020-05-10T10:09:57.5525160Z 
2020-05-10T10:09:57.5525690Z --------- Given  --------
2020-05-10T10:09:57.5525890Z nimble install failed:
2020-05-10T10:09:57.5526590Z $ nimble install -y https://github.com/mratsim/Arraymancer
2020-05-10T10:09:57.5526970Z Downloading https://github.com/mratsim/Arraymancer using git
2020-05-10T10:09:57.5528650Z    Warning: Package 'arraymancer' has an incorrect structure. It should contain a single directory hierarchy for source files, named 'arraymancer', but file 'maxpool2D.nim' is in a directory named 'nn/layers' instead. This will be an error in the future.
2020-05-10T10:09:57.5529790Z       Hint: If 'nn/layers' contains source files for building 'arraymancer', rename it to 'arraymancer'. Otherwise, prevent its installation by adding `skipDirs = @["nn/layers"]` to the .nimble file.
2020-05-10T10:09:57.5529970Z   Verifying dependencies for arraymancer@0.5.2
2020-05-10T10:09:57.5530140Z  Installing nimblas@>= 0.2.2
2020-05-10T10:09:57.5530400Z Downloading https://github.com/unicredit/nimblas using git
2020-05-10T10:09:57.5530590Z   Verifying dependencies for nimblas@0.2.2
2020-05-10T10:09:57.5530740Z  Installing nimblas@0.2.2
2020-05-10T10:09:57.5530870Z    Success: nimblas installed successfully.
2020-05-10T10:09:57.5531130Z  Installing nimlapack@>= 0.1.1
2020-05-10T10:09:57.5531450Z Downloading https://github.com/unicredit/nimlapack using git
2020-05-10T10:09:57.5531660Z   Verifying dependencies for nimlapack@0.2.0
2020-05-10T10:09:57.5531830Z  Installing nimlapack@0.2.0
2020-05-10T10:09:57.5531960Z    Success: nimlapack installed successfully.
2020-05-10T10:09:57.5532120Z  Installing nimcuda@>= 0.1.4
2020-05-10T10:09:57.5532370Z Downloading https://github.com/unicredit/nimcuda using git
2020-05-10T10:09:57.5532560Z   Verifying dependencies for nimcuda@0.1.7
2020-05-10T10:09:57.5532670Z  Installing nimcuda@0.1.7
2020-05-10T10:09:57.5532830Z    Success: nimcuda installed successfully.
2020-05-10T10:09:57.5532980Z  Installing nimcl@>= 0.1.3
2020-05-10T10:09:57.5533220Z Downloading https://github.com/unicredit/nimcl using git
2020-05-10T10:09:57.5533400Z   Verifying dependencies for nimcl@0.1.3
2020-05-10T10:09:57.5533520Z  Installing opencl@>= 1.0
2020-05-10T10:09:57.5534160Z Downloading https://github.com/nim-lang/opencl using git
2020-05-10T10:09:57.5534380Z   Verifying dependencies for opencl@1.0
2020-05-10T10:09:57.5534530Z  Installing opencl@1.0
2020-05-10T10:09:57.5534660Z    Success: opencl installed successfully.
2020-05-10T10:09:57.5534810Z  Installing nimcl@0.1.3
2020-05-10T10:09:57.5534980Z    Success: nimcl installed successfully.
2020-05-10T10:09:57.5535130Z  Installing clblast@any version
2020-05-10T10:09:57.5535760Z Downloading https://github.com/numforge/nim-clblast using git
2020-05-10T10:09:57.5536620Z    Warning: Package 'clblast' has an incorrect structure. It should contain a single directory hierarchy for source files, named 'clblast', but file 'clblast_half.nim' is in a directory named 'generated' instead. This will be an error in the future.
2020-05-10T10:09:57.5537420Z       Hint: If 'generated' contains source files for building 'clblast', rename it to 'clblast'. Otherwise, prevent its installation by adding `skipDirs = @["generated"]` to the .nimble file.
2020-05-10T10:09:57.5537660Z   Verifying dependencies for clblast@0.0.1
2020-05-10T10:09:57.5537860Z       Info: Dependency on opencl@any version already satisfied
2020-05-10T10:09:57.5538030Z   Verifying dependencies for opencl@1.0
2020-05-10T10:09:57.5538180Z  Installing clblast@0.0.1
2020-05-10T10:09:57.5538310Z    Success: clblast installed successfully.
2020-05-10T10:09:57.5538480Z  Installing stb_image@any version
2020-05-10T10:09:57.5539150Z Downloading https://gitlab.com/define-private-public/stb_image-Nim.git using git
2020-05-10T10:09:57.5539430Z download.nim(93)         getTagsListRemote
2020-05-10T10:09:57.5540590Z Error: unhandled exception: Unable to query remote tags for https://gitlab.com/define-private-public/stb_image-Nim.git. Git returned: fatal: unable to access 'https://gitlab.com/define-private-public/stb_image-Nim.git/': The requested URL returned error: 503
2020-05-10T10:09:57.5540810Z  [OSError]
2020-05-10T10:09:57.5540870Z 
2020-05-10T10:09:57.5541360Z -------------------------
2020-05-10T10:09:57.5541440Z 
2020-05-10T10:09:57.5541960Z Category: nimble-packages-1
2020-05-10T10:09:57.5542210Z Name: https://github.com/jackmott/easygl C
2020-05-10T10:09:57.5542670Z Action: run
2020-05-10T10:09:57.5542910Z Result: reInstallFailed
2020-05-10T10:09:57.5543460Z -------- Expected -------
2020-05-10T10:09:57.5543540Z 
2020-05-10T10:09:57.5544040Z --------- Given  --------
2020-05-10T10:09:57.5544450Z nimble install failed:
2020-05-10T10:09:57.5545040Z $ nimble install -y https://github.com/jackmott/easygl
2020-05-10T10:09:57.5545340Z Downloading https://github.com/jackmott/easygl using git
2020-05-10T10:09:57.5545540Z   Verifying dependencies for easygl@0.1.1
2020-05-10T10:09:57.5545690Z  Installing sdl2@any version
2020-05-10T10:09:57.5546310Z Downloading https://github.com/nim-lang/sdl2 using git
2020-05-10T10:09:57.5546470Z   Verifying dependencies for sdl2@2.0.2
2020-05-10T10:09:57.5546620Z  Installing sdl2@2.0.2
2020-05-10T10:09:57.5546780Z    Success: sdl2 installed successfully.
2020-05-10T10:09:57.5546930Z  Installing opengl@any version
2020-05-10T10:09:57.5547500Z Downloading https://github.com/nim-lang/opengl using git
2020-05-10T10:09:57.5547730Z   Verifying dependencies for opengl@1.2.6
2020-05-10T10:09:57.5547870Z  Installing x11@>= 1.1
2020-05-10T10:09:57.5548480Z Downloading https://github.com/nim-lang/x11 using git
2020-05-10T10:09:57.5548670Z   Verifying dependencies for x11@1.1
2020-05-10T10:09:57.5548790Z  Installing x11@1.1
2020-05-10T10:09:57.5548960Z    Success: x11 installed successfully.
2020-05-10T10:09:57.5549110Z  Installing opengl@1.2.6
2020-05-10T10:09:57.5549270Z    Success: opengl installed successfully.
2020-05-10T10:09:57.5549390Z  Installing stb_image@any version
2020-05-10T10:09:57.5550050Z Downloading https://gitlab.com/define-private-public/stb_image-Nim.git using git
2020-05-10T10:09:57.5550330Z download.nim(93)         getTagsListRemote
2020-05-10T10:09:57.5551480Z Error: unhandled exception: Unable to query remote tags for https://gitlab.com/define-private-public/stb_image-Nim.git. Git returned: fatal: unable to access 'https://gitlab.com/define-private-public/stb_image-Nim.git/': The requested URL returned error: 503
2020-05-10T10:09:57.5551720Z  [OSError]
2020-05-10T10:09:57.5551780Z 
2020-05-10T10:09:57.5552280Z -------------------------
2020-05-10T10:09:57.5552410Z 
2020-05-10T10:09:57.5552870Z Category: nimble-packages-1
2020-05-10T10:09:57.5553140Z Name: https://github.com/treeform/fidget C
2020-05-10T10:09:57.5553310Z Action: run
2020-05-10T10:09:57.5553490Z Result: reInstallFailed
2020-05-10T10:09:57.5554000Z -------- Expected -------
2020-05-10T10:09:57.5554080Z 
2020-05-10T10:09:57.5554560Z --------- Given  --------
2020-05-10T10:09:57.5554760Z nimble install failed:
2020-05-10T10:09:57.5555360Z $ nimble install -y https://github.com/treeform/fidget
2020-05-10T10:09:57.5555600Z Downloading https://github.com/treeform/fidget using git
2020-05-10T10:09:57.5555800Z   Verifying dependencies for fidget@0.3.2
2020-05-10T10:09:57.5555960Z  Installing chroma@>= 0.0.1
2020-05-10T10:09:57.5556200Z Downloading https://github.com/treeform/chroma using git
2020-05-10T10:09:57.5556390Z   Verifying dependencies for chroma@0.1.0
2020-05-10T10:09:57.5556500Z  Installing chroma@0.1.0
2020-05-10T10:09:57.5556670Z    Success: chroma installed successfully.
2020-05-10T10:09:57.5556820Z  Installing typography@>= 0.2.2
2020-05-10T10:09:57.5557070Z Downloading https://github.com/treeform/typography using git
2020-05-10T10:09:57.5557270Z   Verifying dependencies for typography@0.3.0
2020-05-10T10:09:57.5557390Z  Installing flippy@>= 0.4.2
2020-05-10T10:09:57.5557640Z Downloading https://github.com/treeform/flippy using git
2020-05-10T10:09:57.5557830Z   Verifying dependencies for flippy@0.4.4
2020-05-10T10:09:57.5558060Z  Installing vmath@>= 0.1.0
2020-05-10T10:09:57.5558310Z Downloading https://github.com/treeform/vmath using git
2020-05-10T10:09:57.5558500Z   Verifying dependencies for vmath@0.3.1
2020-05-10T10:09:57.5558720Z  Installing vmath@0.3.1
2020-05-10T10:09:57.5558850Z    Success: vmath installed successfully.
2020-05-10T10:09:57.5559040Z       Info: Dependency on chroma@>= 0.0.1 already satisfied
2020-05-10T10:09:57.5559450Z   Verifying dependencies for chroma@0.1.0
2020-05-10T10:09:57.5559750Z  Installing print@>= 0.1.0
2020-05-10T10:09:57.5559990Z Downloading https://github.com/treeform/print using git
2020-05-10T10:09:57.5560240Z   Verifying dependencies for print@0.1.0
2020-05-10T10:09:57.5560660Z  Installing print@0.1.0
2020-05-10T10:09:57.5560820Z    Success: print installed successfully.
2020-05-10T10:09:57.5560970Z  Installing nimPNG@>= 0.2.6
2020-05-10T10:09:57.5561180Z Downloading https://github.com/jangko/nimPNG using git
2020-05-10T10:09:57.5561370Z   Verifying dependencies for nimPNG@0.2.6
2020-05-10T10:09:57.5561520Z  Installing nimPNG@0.2.6
2020-05-10T10:09:57.5561690Z    Success: nimPNG installed successfully.
2020-05-10T10:09:57.5561800Z  Installing snappy@>= 0.1.0
2020-05-10T10:09:57.5562050Z Downloading https://github.com/jangko/snappy using git
2020-05-10T10:09:57.5562240Z   Verifying dependencies for snappy@0.1.0
2020-05-10T10:09:57.5562380Z  Installing snappy@0.1.0
2020-05-10T10:09:57.5562550Z    Success: snappy installed successfully.
2020-05-10T10:09:57.5562670Z  Installing flippy@0.4.4
2020-05-10T10:09:57.5562830Z    Success: flippy installed successfully.
2020-05-10T10:09:57.5563010Z       Info: Dependency on vmath@>= 0.2.0 already satisfied
2020-05-10T10:09:57.5563180Z   Verifying dependencies for vmath@0.3.1
2020-05-10T10:09:57.5563330Z       Info: Dependency on chroma@>= 0.0.1 already satisfied
2020-05-10T10:09:57.5563490Z   Verifying dependencies for chroma@0.1.0
2020-05-10T10:09:57.5563680Z       Info: Dependency on print@>= 0.1.0 already satisfied
2020-05-10T10:09:57.5563840Z   Verifying dependencies for print@0.1.0
2020-05-10T10:09:57.5563990Z  Installing typography@0.3.0
2020-05-10T10:09:57.5564130Z    Success: typography installed successfully.
2020-05-10T10:09:57.5564310Z       Info: Dependency on flippy@>= 0.2.0 already satisfied
2020-05-10T10:09:57.5564470Z   Verifying dependencies for flippy@0.4.4
2020-05-10T10:09:57.5564660Z       Info: Dependency on vmath@>= 0.1.0 already satisfied
2020-05-10T10:09:57.5564790Z   Verifying dependencies for vmath@0.3.1
2020-05-10T10:09:57.5564970Z       Info: Dependency on chroma@>= 0.0.1 already satisfied
2020-05-10T10:09:57.5565140Z   Verifying dependencies for chroma@0.1.0
2020-05-10T10:09:57.5565330Z       Info: Dependency on print@>= 0.1.0 already satisfied
2020-05-10T10:09:57.5565490Z   Verifying dependencies for print@0.1.0
2020-05-10T10:09:57.5565640Z       Info: Dependency on nimPNG@>= 0.2.6 already satisfied
2020-05-10T10:09:57.5565810Z   Verifying dependencies for nimPNG@0.2.6
2020-05-10T10:09:57.5565990Z       Info: Dependency on snappy@>= 0.1.0 already satisfied
2020-05-10T10:09:57.5566150Z   Verifying dependencies for snappy@0.1.0
2020-05-10T10:09:57.5566300Z       Info: Dependency on vmath@>= 0.3.0 already satisfied
2020-05-10T10:09:57.5566470Z   Verifying dependencies for vmath@0.3.1
2020-05-10T10:09:57.5566650Z       Info: Dependency on print@>= 0.1.0 already satisfied
2020-05-10T10:09:57.5566810Z   Verifying dependencies for print@0.1.0
2020-05-10T10:09:57.5566960Z       Info: Dependency on opengl@>= 1.2.3 already satisfied
2020-05-10T10:09:57.5567140Z   Verifying dependencies for opengl@1.2.6
2020-05-10T10:09:57.5567320Z       Info: Dependency on x11@>= 1.1 already satisfied
2020-05-10T10:09:57.5567480Z   Verifying dependencies for x11@1.1
2020-05-10T10:09:57.5567640Z  Installing html5_canvas@>= 1.3
2020-05-10T10:09:57.5568620Z Downloading https://gitlab.com/define-private-public/HTML5-Canvas-Nim using git
2020-05-10T10:09:57.5568930Z download.nim(93)         getTagsListRemote
2020-05-10T10:09:57.5570080Z Error: unhandled exception: Unable to query remote tags for https://gitlab.com/define-private-public/HTML5-Canvas-Nim. Git returned: fatal: unable to access 'https://gitlab.com/define-private-public/HTML5-Canvas-Nim/': The requested URL returned error: 503
2020-05-10T10:09:57.5570310Z  [OSError]
2020-05-10T10:09:57.5570370Z 
2020-05-10T10:09:57.5571400Z -------------------------
2020-05-10T10:09:57.5571620Z 
2020-05-10T10:09:57.5572180Z Category: nimble-packages-1
2020-05-10T10:09:57.5572780Z Name: https://github.com/Vindaar/ggplotnim C
2020-05-10T10:09:57.5573010Z Action: run
2020-05-10T10:09:57.5573190Z Result: reInstallFailed
2020-05-10T10:09:57.5573730Z -------- Expected -------
2020-05-10T10:09:57.5574000Z 
2020-05-10T10:09:57.5574520Z --------- Given  --------
2020-05-10T10:09:57.5574720Z nimble install failed:
2020-05-10T10:09:57.5575320Z $ nimble install -y https://github.com/Vindaar/ggplotnim
2020-05-10T10:09:57.5575570Z Downloading https://github.com/Vindaar/ggplotnim using git
2020-05-10T10:09:57.5575760Z   Verifying dependencies for ggplotnim@0.3.5
2020-05-10T10:09:57.5575940Z  Installing https://github.com/Vindaar/seqmath@>= 0.1.7
2020-05-10T10:09:57.5576190Z Downloading https://github.com/Vindaar/seqmath using git
2020-05-10T10:09:57.5576380Z   Verifying dependencies for seqmath@0.1.10
2020-05-10T10:09:57.5576490Z  Installing seqmath@0.1.10
2020-05-10T10:09:57.5576720Z    Success: seqmath installed successfully.
2020-05-10T10:09:57.5576920Z  Installing ginger@>= 0.2.3
2020-05-10T10:09:57.5577180Z Downloading https://github.com/Vindaar/ginger using git
2020-05-10T10:09:57.5577340Z   Verifying dependencies for ginger@0.2.3
2020-05-10T10:09:57.5577520Z       Info: Dependency on chroma@>= 0.1.0 already satisfied
2020-05-10T10:09:57.5577690Z   Verifying dependencies for chroma@0.1.0
2020-05-10T10:09:57.5577910Z       Info: Dependency on https://github.com/vindaar/seqmath@>= 0.1.7 already satisfied
2020-05-10T10:09:57.5578090Z   Verifying dependencies for seqmath@0.1.10
2020-05-10T10:09:57.5578200Z  Installing cairo@any version
2020-05-10T10:09:57.5578800Z Downloading https://github.com/nim-lang/cairo using git
2020-05-10T10:09:57.5579670Z    Warning: Package 'cairo' has an incorrect structure. The top level of the package source directory should contain at most one module, named 'cairo.nim', but a file named 'cairowin32.nim' was found. This will be an error in the future.
2020-05-10T10:09:57.5580950Z       Hint: If this is the primary source file in the package, rename it to 'cairo.nim'. If it's a source file required by the main module, or if it is one of several modules exposed by 'cairo', then move it into a 'cairo/' subdirectory. If it's a test file or otherwise not required to build the the package 'cairo.nim', prevent its installation by adding `skipFiles = @["cairowin32.nim"]` to the .nimble file. See https://github.com/nim-lang/nimble#libraries for more info.
2020-05-10T10:09:57.5581220Z   Verifying dependencies for cairo@1.1.1
2020-05-10T10:09:57.5581360Z  Installing cairo@1.1.1
2020-05-10T10:09:57.5597590Z    Success: cairo installed successfully.
2020-05-10T10:09:57.5597810Z  Installing ginger@0.2.3
2020-05-10T10:09:57.5597950Z    Success: ginger installed successfully.
2020-05-10T10:09:57.5598080Z  Installing persvector@>= 1.0.0
2020-05-10T10:09:57.5598900Z Downloading https://github.com/PMunch/nim-persistent-vector using git
2020-05-10T10:09:57.5599110Z   Verifying dependencies for persvector@1.0.0
2020-05-10T10:09:57.5599260Z  Installing persvector@1.0.0
2020-05-10T10:09:57.5599410Z    Success: persvector installed successfully.
2020-05-10T10:09:57.5599560Z  Installing shell@>= 0.2.2
2020-05-10T10:09:57.5599800Z Downloading https://github.com/Vindaar/shell using git
2020-05-10T10:09:57.5599990Z   Verifying dependencies for shell@0.4.0
2020-05-10T10:09:57.5600110Z  Installing shell@0.4.0
2020-05-10T10:09:57.5600270Z    Success: shell installed successfully.
2020-05-10T10:09:57.5600420Z  Installing arraymancer@any version
2020-05-10T10:09:57.5600680Z Downloading https://github.com/mratsim/Arraymancer using git
2020-05-10T10:09:57.5601660Z    Warning: Package 'arraymancer' has an incorrect structure. It should contain a single directory hierarchy for source files, named 'arraymancer', but file 'maxpool2D.nim' is in a directory named 'nn/layers' instead. This will be an error in the future.
2020-05-10T10:09:57.5602770Z       Hint: If 'nn/layers' contains source files for building 'arraymancer', rename it to 'arraymancer'. Otherwise, prevent its installation by adding `skipDirs = @["nn/layers"]` to the .nimble file.
2020-05-10T10:09:57.5603020Z   Verifying dependencies for arraymancer@0.5.2
2020-05-10T10:09:57.5603180Z       Info: Dependency on nimblas@>= 0.2.2 already satisfied
2020-05-10T10:09:57.5603550Z   Verifying dependencies for nimblas@0.2.2
2020-05-10T10:09:57.5603730Z       Info: Dependency on nimlapack@>= 0.1.1 already satisfied
2020-05-10T10:09:57.5603900Z   Verifying dependencies for nimlapack@0.2.0
2020-05-10T10:09:57.5604050Z       Info: Dependency on nimcuda@>= 0.1.4 already satisfied
2020-05-10T10:09:57.5604210Z   Verifying dependencies for nimcuda@0.1.7
2020-05-10T10:09:57.5604400Z       Info: Dependency on nimcl@>= 0.1.3 already satisfied
2020-05-10T10:09:57.5604560Z   Verifying dependencies for nimcl@0.1.3
2020-05-10T10:09:57.5604840Z       Info: Dependency on opencl@>= 1.0 already satisfied
2020-05-10T10:09:57.5604990Z   Verifying dependencies for opencl@1.0
2020-05-10T10:09:57.5605250Z       Info: Dependency on clblast@any version already satisfied
2020-05-10T10:09:57.5605430Z   Verifying dependencies for clblast@0.0.1
2020-05-10T10:09:57.5605620Z       Info: Dependency on opencl@any version already satisfied
2020-05-10T10:09:57.5605750Z   Verifying dependencies for opencl@1.0
2020-05-10T10:09:57.5605910Z  Installing stb_image@any version
2020-05-10T10:09:57.5606640Z Downloading https://gitlab.com/define-private-public/stb_image-Nim.git using git
2020-05-10T10:09:57.5606910Z download.nim(93)         getTagsListRemote
2020-05-10T10:09:57.5608090Z Error: unhandled exception: Unable to query remote tags for https://gitlab.com/define-private-public/stb_image-Nim.git. Git returned: fatal: unable to access 'https://gitlab.com/define-private-public/stb_image-Nim.git/': The requested URL returned error: 503
2020-05-10T10:09:57.5608320Z  [OSError]
2020-05-10T10:09:57.5608380Z 
2020-05-10T10:09:57.5608890Z -------------------------
2020-05-10T10:09:57.5608960Z 
2020-05-10T10:09:57.5609470Z Category: nimble-packages-1
2020-05-10T10:09:57.5609680Z Name: PackageFileParsed C
2020-05-10T10:09:57.5609840Z Action: run
2020-05-10T10:09:57.5610010Z Result: reBuildFailed
2020-05-10T10:09:57.5610470Z -------- Expected -------
2020-05-10T10:09:57.5610590Z 
2020-05-10T10:09:57.5611100Z --------- Given  --------
2020-05-10T10:09:57.5611170Z 
2020-05-10T10:09:57.5611670Z -------------------------
```

--- a/lib/std/private/nimbleutils.nim
+++ b/lib/std/private/nimbleutils.nim
@@ -1,0 +1,32 @@
+##[
+internal API for now, API subject to change
+]##
+
+import std/[os,osproc,sugar,strutils]
+
+proc actionRetry*(maxRetry: int, backoffDuration: float, action: proc(): bool): bool =
+  ## retry `action` up to `maxRetry` times with exponential backoff and initial
+  ## duraton of `backoffDuration` seconds
+  var t = backoffDuration
+  for i in 0..<maxRetry:
+    if action(): return true
+    if i == maxRetry - 1: break
+    sleep(int(t * 1000))
+    t = t * 2 # exponential backoff
+  return false
+
+proc nimbleInstall*(name: string, message: var string): bool =
+  let cmd = "nimble install -y " & name
+  let (outp, status) = execCmdEx(cmd)
+  if status != 0:
+    message = "'$1' failed:\n$2" % [cmd, outp]
+    result = false
+  else: result = true
+
+when isMainModule:
+  block:
+    var msg: string
+    let ok = actionRetry(maxRetry = 2, backoffDuration = 0.1):
+      (proc(): bool = nimbleInstall("nonexistant", msg))
+    doAssert "Package not found" in msg
+    doAssert not ok

--- a/testament/categories.nim
+++ b/testament/categories.nim
@@ -13,6 +13,7 @@
 # included from testament.nim
 
 import important_packages
+import std/private/nimbleutils
 
 const
   specialCategories = [
@@ -501,9 +502,9 @@ proc testNimblePackages(r: var TResults; cat: Category; packageFilter: string, p
       if not existsDir(buildPath):
         if hasDep:
           let installName = if url.len != 0: url else: name
-          let (nimbleCmdLine, nimbleOutput, nimbleStatus) = execCmdEx2("nimble", ["install", "-y", installName])
-          if nimbleStatus != QuitSuccess:
-            let message = "nimble install failed:\n$ " & nimbleCmdLine & "\n" & nimbleOutput
+          var message: string
+          if not actionRetry(maxRetry = 3, backoffDuration = 1.0,
+            (proc(): bool = nimbleInstall(installName, message))):
             r.addResult(test, targetC, "", message, reInstallFailed)
             continue
 

